### PR TITLE
feat(zk): improve performance of zk pke proofs

### PR DESCRIPTION
**Pull request by sarah-ek**
closes: _please link all relevant issues_

### Check-list:

* [x] Tests for the changes have been added (for bug fixes / features)
* [x] Docs have been added / updated (for bug fixes / features)
* [ ] Relevant issues are marked as resolved/closed, related issues are linked in the description
* [ ] Check for breaking changes (including serialization changes) and add them to commit message following the conventional commit [specification][conventional-breaking]

[conventional-breaking]: https://www.conventionalcommits.org/en/v1.0.0/#commit-message-with-description-and-breaking-change-footer

### CODE REVIEW

File: tfhe-zk-pok/Cargo.toml [hunk 1]
```diff
@@ -12,15 +12,15 @@ description = "tfhe-zk-pok: An implementation of zero-knowledge proofs of encryp
 # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
 
 [dependencies]
-ark-bls12-381 = "0.4.0"
-ark-ec = "0.4.2"
-ark-ff = "0.4.2"
-ark-poly = "0.4.2"
+ark-bls12-381 = { package = "tfhe-ark-bls12-381", version = "0.4.0" }
+ark-ec = { package = "tfhe-ark-ec", version = "0.4.2" }
+ark-ff = { package = "tfhe-ark-ff", version = "0.4.2" }
```
Comment (IceTDrinker) [inline tfhe-zk-pok/Cargo.toml:17]: can we have 0.4.3 to make sure we have the patched one ? :)

File: tfhe-zk-pok/src/curve_api.rs [hunk 2]
```diff
@@ -198,19 +224,32 @@ impl CurveGroupOps<bls12_381::Zp> for bls12_381::G1 {
     fn double(self) -> Self {
         self.double()
     }
+
+    fn normalize(self) -> Self::Affine {
```
Comment (IceTDrinker) [inline tfhe-zk-pok/src/curve_api.rs:228]: normalization is the act of getting the affine representation of a projective element ?

File: tfhe-zk-pok/src/curve_api.rs [hunk 3]
```diff
@@ -164,6 +180,9 @@ impl FieldOps for bls12_381::Zp {
     fn hash(values: &mut [Self], data: &[&[u8]]) {
         Self::hash(values, data)
     }
+    fn hash_128bit(values: &mut [Self], data: &[&[u8]]) {
```
Comment (IceTDrinker) [inline tfhe-zk-pok/src/curve_api.rs:183]: why the adition ?

File: tfhe-zk-pok/src/curve_api/bls12_381.rs [hunk 4]
```diff
@@ -36,6 +36,38 @@ fn bigint_to_bytes(x: [u64; 6]) -> [u8; 6 * 8] {
...
+    impl G1Affine {
+        pub fn multi_mul_scalar(bases: &[Self], scalars: &[Zp]) -> G1 {
+            G1 {
+                inner: ark_bls12_381::g1::G1Projective::msm(
+                    unsafe {
+                        &*(bases as *const [G1Affine] as *const [ark_bls12_381::g1::G1Affine])
+                    },
+                    unsafe { &*(scalars as *const [Zp] as *const [ark_bls12_381::Fr]) },
+                )
+                .unwrap(),
```
Comment (IceTDrinker) [inline tfhe-zk-pok/src/curve_api/bls12_381.rs:66]: why the need for unsafe ? and what are the safety requirements ?

File: tfhe-zk-pok/src/curve_api/bls12_381.rs [hunk 5]
```diff
@@ -179,6 +211,38 @@ mod g1 {
...
+    }
+
+    impl G2Affine {
+        pub fn multi_mul_scalar(bases: &[Self], scalars: &[Zp]) -> G2 {
+            G2 {
+                inner: ark_bls12_381::g2::G2Projective::msm(
+                    unsafe {
+                        &*(bases as *const [G2Affine] as *const [ark_bls12_381::g2::G2Affine])
+                    },
+                    unsafe { &*(scalars as *const [Zp] as *const [ark_bls12_381::Fr]) },
```
Comment (IceTDrinker) [inline tfhe-zk-pok/src/curve_api/bls12_381.rs:239]: same question

File: tfhe-zk-pok/src/proofs/rlwe.rs [hunk 6]
```diff
@@ -763,7 +829,11 @@ pub fn verify<G: Curve>(
         if e(pi, G::G2::GENERATOR)
             != e(c_y.mul_scalar(delta_y) + c_h, c_hat)
                 - e(c_y.mul_scalar(delta_eq), c_hat_t)
-                - e(g_list[1], g_hat_list[n]).mul_scalar(t_theta * delta_theta)
+                - e(
+                    G::G1::projective(g_list[1]),
```
Comment (IceTDrinker) [inline tfhe-zk-pok/src/proofs/rlwe.rs:833]: some comment on the use of projective could be useful :)

File: tfhe-zk-pok/src/proofs/mod.rs [hunk 7]
```diff
@@ -127,6 +128,8 @@ impl<G: Curve> GroupElements<G> {
     }
 }
 
+pub const HASH_METADATA_LEN_BYTES: usize = 256;
```
Comment (IceTDrinker) [inline tfhe-zk-pok/src/proofs/mod.rs:131]: is that required for the performance ?

If not then I would have that in a separate PR please

**Review by IceTDrinker — COMMENTED** [review]
A few general questions to start, notably on the new hash arrays with metadata

File: tfhe-zk-pok/Cargo.toml (hunk 1 above)
Comment (sarah-ek) [reply tfhe-zk-pok/Cargo.toml:17]: okay

File: tfhe-zk-pok/src/curve_api.rs (hunk 2 above)
Comment (sarah-ek) [reply tfhe-zk-pok/src/curve_api.rs:228]: yes

File: tfhe-zk-pok/src/curve_api.rs (hunk 3 above)
Comment (sarah-ek) [reply tfhe-zk-pok/src/curve_api.rs:183]: because according to Benoit we could use a 128 bit exponent for the `t_i`, so i added a new hash function that generates those

File: tfhe-zk-pok/src/curve_api/bls12_381.rs (hunk 4 above)
Comment (sarah-ek) [reply tfhe-zk-pok/src/curve_api/bls12_381.rs:66]: this is essentially interpreting the wrapper struct `Zp(ark_bls12_381::Fr)` as its contents `Fr`. this is safe since it's `repr(transparent)`. we can remove the unsafety but that would require an allocation + copy which has non negligible cost, at least for the bases

File: tfhe-zk-pok/src/proofs/rlwe.rs (hunk 6 above)
Comment (sarah-ek) [reply tfhe-zk-pok/src/proofs/rlwe.rs:771]: this is because the lists are stored in affine representation, since that's the faster one for multiexponentiation, but the operations require the projective representation. conversion from affine to projective is effectively a no-op

File: tfhe-zk-pok/src/proofs/mod.rs (hunk 7 above)
Comment (sarah-ek) [reply tfhe-zk-pok/src/proofs/mod.rs]: will do

File: tfhe-zk-pok/src/curve_api/bls12_381.rs (hunk 4 above)
Comment (IceTDrinker) [reply tfhe-zk-pok/src/curve_api/bls12_381.rs:66]: I'm fine with the unsafe as long as the comment about the repr transparent is given in the code :)

Comment (IceTDrinker) [reply tfhe-zk-pok/src/curve_api/bls12_381.rs:66]: also if the wrapper proves to be an issue it can be removed down the line, the BLS446 is the final choice now

File: tfhe-zk-pok/src/proofs/rlwe.rs (hunk 6 above)
Comment (IceTDrinker) [reply tfhe-zk-pok/src/proofs/rlwe.rs:771]: ok no op is fine by me !

File: tfhe-zk-pok/src/curve_api/bls12_381.rs (hunk 4 above)
Comment (sarah-ek) [reply tfhe-zk-pok/src/curve_api/bls12_381.rs:67]: added a comment `// SAFETY: interpreting a `repr(transparent)` pointer as its contents.`

File: tfhe-zk-pok/src/curve_api.rs (hunk 3 above)
Comment (IceTDrinker) [reply tfhe-zk-pok/src/curve_api.rs:183]: is it used currently ?

File: tfhe-zk-pok/src/curve_api/bls12_381.rs (hunk 4 above)
Comment (IceTDrinker) [reply tfhe-zk-pok/src/curve_api/bls12_381.rs:67]: thanks !

File: tfhe-zk-pok/src/proofs/pke.rs [hunk 8]
```diff
@@ -192,6 +195,7 @@ pub fn prove<G: Curve>(
 
     assert!(c2.len() <= k);
     let k = k.min(c2.len());
+    let (_, big_d, _) = compute_crs_params(d, k, b_i, q, t);
```
Comment (IceTDrinker) [inline tfhe-zk-pok/src/proofs/pke.rs:198]: why this change ? 🤔

File: tfhe-zk-pok/src/proofs/pke.rs [hunk 9]
```diff
@@ -726,7 +737,7 @@ pub fn verify<G: Curve>(
     }
 
     let mut t = vec![G::Zp::ZERO; n];
-    G::Zp::hash(
+    G::Zp::hash_128bit(
```
Comment (IceTDrinker) [inline tfhe-zk-pok/src/proofs/pke.rs:740]: is it from the original ZK formulation ?

File: tfhe-zk-pok/src/proofs/pke.rs [hunk 10]
```diff
@@ -822,13 +838,17 @@ pub fn verify<G: Curve>(
...
+                pi_kzg,
+                G::G2::projective(g_hat_list[1]) - G::G2::GENERATOR.mul_scalar(z),
+            )
         {
             Ok(())
         } else {
             Err(())
         }
     } else {
+        // PERF: rewrite as multi_mul_scalar?
```
Comment (IceTDrinker) [inline tfhe-zk-pok/src/proofs/pke.rs:851]: should it be done ?

File: tfhe-zk-pok/src/curve_api.rs (hunk 3 above)
Comment (sarah-ek) [reply tfhe-zk-pok/src/curve_api.rs:183]: yeah, it's used for the pke proof

File: tfhe-zk-pok/src/proofs/pke.rs (hunk 9 above)
Comment (sarah-ek) [reply tfhe-zk-pok/src/proofs/pke.rs:740]: this is from Benoit's remark about optimizing the G2 multi exp

File: tfhe-zk-pok/src/proofs/pke.rs (hunk 10 above)
Comment (sarah-ek) [reply tfhe-zk-pok/src/proofs/pke.rs:851]: yeah, but it's not urgent since it's for the slow verify, which isn't used at the moment afaik

Comment (IceTDrinker) [reply tfhe-zk-pok/src/proofs/pke.rs:851]: ok I'll create an issue for it then

Comment (IceTDrinker) [reply tfhe-zk-pok/src/proofs/pke.rs:851]: https://github.com/zama-ai/tfhe-rs-internal/issues/616

File: tfhe-zk-pok/src/proofs/pke.rs (hunk 9 above)
Comment (IceTDrinker) [reply tfhe-zk-pok/src/proofs/pke.rs:740]: ok !

File: tfhe-zk-pok/src/proofs/pke.rs (hunk 8 above)
Comment (sarah-ek) [reply tfhe-zk-pok/src/proofs/pke.rs:198]: i think i was experimenting with using a different value of D when using a smaller `k`, but forgot to remove it. would you like me to undo it?

Comment (IceTDrinker) [reply tfhe-zk-pok/src/proofs/pke.rs:198]: if the change is not required yes please

**Review by IceTDrinker — APPROVED** [review]
Thanks a lot !
